### PR TITLE
Allow IntraL0 compaction in FIFO Compaction

### PR DESCRIPTION
**Pull request by siying**
Summary: Allow an option for users to do some compaction in FIFO compaction, to pay some write amplification for fewer number of files.

Test Plan: Add a unit test

### CODE REVIEW

**Comment by siying** [comment]
Some benchmark results:

Without the change:

[sdong@dev12075.prn1 /data/users/sdong/rocksdb7] TEST_TMPDIR=/dev/shm/ ./db_bench --benchmarks=readwhilewriting -compaction_style=2 --num=20000000 -level0_file_num_compaction_trigger=8 -bloom_bits=10 --threads=16

readwhilewriting :       0.725 micros/op 1378846 ops/sec;   54.7 MB/s (7070947 of 20000000 found)

With the change:

] TEST_TMPDIR=/dev/shm/ ./db_bench --benchmarks=readwhilewriting -compaction_style=2 --num=20000000 -level0_file_num_compaction_trigger=8 -bloom_bits=10 --threads=16 --fifo_compaction_allow_compaction --fifo_compaction_max_table_files_size_mb=1000

readwhilewriting :       0.485 micros/op 2062525 ops/sec;   78.0 MB/s (6766309 of 20000000 found)

With the change, the SST files under the directory after the compaction looks like this:

-rw-r--r--. 1 sdong users 241354289 Apr 14 17:45 000108.sst
-rw-r--r--. 1 sdong users 241391023 Apr 14 17:45 000125.sst
-rw-r--r--. 1 sdong users 241431557 Apr 14 17:45 000142.sst
-rw-r--r--. 1 sdong users  32490220 Apr 14 17:45 000144.sst
-rw-r--r--. 1 sdong users  32478382 Apr 14 17:45 000146.sst
-rw-r--r--. 1 sdong users  32498268 Apr 14 17:45 000148.sst

**Comment by ajkr** [comment]
just curious what was the motivation to introduce this for FIFO compaction?

**Comment by siying** [comment]
@ajkr a user sees high CPU overhead when there are many files in FIFO. I just want to create a solution for them to try.

File: include/rocksdb/advanced_options.h [hunk 1]
```diff
@@ -62,9 +62,19 @@ struct CompactionOptionsFIFO {
   // Default: 1GB
   uint64_t max_table_files_size;
 
+  // If true, try to do compaction to compact smaller files into larger ones.
+  // Minimum files to compact follows options.level0_file_num_compaction_trigger
+  // and compaction won't trigger if average compact bytes per del file is
+  // larger than options.write_buffer_size * 2. This is to protect large files
```
Comment (ajkr) [inline include/rocksdb/advanced_options.h]: looks like the "* 2" is no longer the case

**Pull request closed without merging**